src/bm_sim/periodic_task.cpp

```diff
@@ -0,0 +1,170 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bm/bm_sim/periodic_task.h>
+#include <bm/bm_sim/logger.h>
+
+#include <string>
+#include <iostream>
+
+namespace bm {
+
+PeriodicTask::PeriodicTask(
+    const std::string &name, std::function<void()> fn,
+    std::chrono::milliseconds interval)
+    : name(name), interval(interval),
+      next(std::chrono::system_clock::now() + interval), fn(fn) {
+  PeriodicTaskList::get_instance().register_task(this);
+}
+
+PeriodicTask::~PeriodicTask() {
+  // Destructor automatically unregisters the task
+  cancel();
+}
+
+void
+PeriodicTask::cancel() {
+  PeriodicTaskList::get_instance().unregister_task(this);
+}
+
+void
+PeriodicTask::reset_next() {
+  next = std::chrono::system_clock::now() + interval;
+}
+
+void
+PeriodicTask::execute() {
+  fn();
+  reset_next();
+}
+
+constexpr std::chrono::milliseconds PeriodicTaskList::kDefaultTimeout;
+
+bool
+PeriodicTaskList::contains_task(PeriodicTask *task) {
+  std::lock_guard<std::mutex> lock(queue_mutex);
+
+  TaskQueue tmp;
```

Comment:
This is why it probably makes sense to differentiate between a *registered task* and the *next execution of a task*
You could create a new `struct` for *next execution of task* and a new data structure mapping task pointer to *next execution of task* pointer. This would let you check in constant time if a task exists and would also let you do constant-time "lazy" removal of a task.
However, performance is not really a concern for task registration & removal so I won't block the PR for this.